src/compiler/checker.ts

```diff
@@ -2846,7 +2982,14 @@ namespace ts {
             }
 
             function signatureToSignatureDeclarationHelper(signature: Signature, kind: SyntaxKind, context: NodeBuilderContext): SignatureDeclaration {
-                const typeParameters = signature.typeParameters && signature.typeParameters.map(parameter => typeParameterToDeclaration(parameter, context));
+                let typeParameters: TypeParameterDeclaration[];
+                let typeArguments: TypeNode[];
+                if (context.flags & NodeBuilderFlags.WriteTypeArgumentsOfSignature && signature.target && signature.mapper && signature.target.typeParameters) {
+                    typeArguments = signature.target.typeParameters.map(parameter => typeToTypeNodeHelper(instantiateType(parameter, signature.mapper), context));

```

Comment:
why not instantiate the whole signature and call this function recursively? Would that call be missing context of some kind?

Reply:
When we instantiate a signature, we instantiate the parameter types and the return type - type parameters get copied to new copies of themselves, but are not instantiated with the mapper until needed (hence why it needs to be handled on the consuming end). If we did instantiate them, they would no longer be guaranteed to be `TypeParameter`s, and would no longer be able to be stored in the signature's `typeParameter` array. 😉